src/Zynga/Framework/StorableObject/Collections/Map/V2/Base.hh

```diff
@@ -0,0 +1,188 @@
+<?hh // strict
+
+namespace Zynga\Framework\StorableObject\Collections\Map\V2;
+
+use
+  Zynga\Framework\StorableObject\Collections\Map\V2\Exporter\Base as Exporter
+;
+use
+  Zynga\Framework\StorableObject\Collections\Map\V2\Importer\Box as BoxImporter
+;
+use
+  Zynga\Framework\StorableObject\Collections\Map\V2\Importer\Storable as StorableImporter
+;
+use
+  Zynga\Framework\StorableObject\Collections\V2\Interfaces\StorableMapCollection
+;
+use
+  Zynga\Framework\StorableObject\V1\Exceptions\OperationNotSupportedException
+;
+use Zynga\Framework\StorableObject\V1\Fields;
+use Zynga\Framework\StorableObject\V1\Fields\Generic as FieldsGeneric;
+use Zynga\Framework\StorableObject\V1\Interfaces\ExportInterface;
+use Zynga\Framework\StorableObject\V1\Interfaces\FieldsInterface;
+use Zynga\Framework\StorableObject\V1\Interfaces\ImportInterface;
+use Zynga\Framework\StorableObject\V1\Interfaces\StorableObjectInterface;
+use Zynga\Framework\Type\V1\Interfaces\TypeInterface;
+
+class Base<Tv> implements StorableMapCollection<Tv> {
+  /**
+   * Native Map only supports int keys or string keys
+   * https://docs.hhvm.com/hack/reference/class/HH.Map/
+   * We are forcing it to support only string as for data with
+   * int keys we can make use of StorableVector
+   */
+  private Map<string, Tv> $map;
+  private bool $isRequired;
+  private bool $isDefaultValue;
+
+  public function __construct(private classname<Tv> $valueType) {
+    $interfaces = class_implements($valueType);
+    if (array_key_exists(StorableObjectInterface::class, $interfaces) === false &&
```

Comment:
Should we move this check to a central Collection helper class or maybe a trait? The reason I bring that up is because this check would be consistent across all collections.

So something like

```
// Check out how we do traits with: https://docs.hhvm.com/hack/other-features/trait-and-interface-requirements
trait CollectionTrait {
  public function typeImplementsRequiredInterface<Type>(private classname<Type> $rawType) {

    $interfaces = class_implements($rawType);
    if (array_key_exists(StorableObjectInterface::class, $interfaces) === false &&
        array_key_exists(TypeInterface::class, $interfaces) === false) {
      throw new OperationNotSupportedException(
        'Collection only support the following types:'.
        StorableObjectInterface::class.
        ', '.
        TypeInterface::class,
      );
    }
  }
}

// Then in your class and others

class Map<Value> implements //...  {
   use CollectionTrait;
}
```